nav2_costmap_2d/test/unit/keepout_filter_test.cpp

```diff
@@ -0,0 +1,363 @@
+      std::bind(&InfoPublisher::infoCallback, this, std::placeholders::_1));
+  }
+
+  ~InfoPublisher()
+  {
+    subscription_.reset();
+    publisher_.reset();
+    delete access_;
+  }
+
+  bool isPublished()
+  {
+    std::lock_guard<mutex_t> guard(*getMutex());
+    return is_published_;
+  }
+
+  mutex_t * getMutex()
+  {
+    return access_;
+  }
+
+private:
+  rclcpp::Publisher<nav2_msgs::msg::CostmapFilterInfo>::SharedPtr publisher_;
+  rclcpp::Subscription<nav2_msgs::msg::CostmapFilterInfo>::SharedPtr subscription_;
+
+  void infoCallback(const nav2_msgs::msg::CostmapFilterInfo::SharedPtr)
+  {
+    std::lock_guard<mutex_t> guard(*getMutex());
+    is_published_ = true;
+  }
+
+  bool is_published_;
+
+  mutex_t * access_;
+};  // InfoPublisher
+
+class MaskPublisher : public rclcpp::Node
+{
+public:
+  MaskPublisher(const nav_msgs::msg::OccupancyGrid & mask)
+  : Node("mask_pub"), is_published_(false)
+  {
+    access_ = new mutex_t();
+
+    publisher_ = this->create_publisher<nav_msgs::msg::OccupancyGrid>(
+      MASK_TOPIC,
+      rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable());
+
+    publisher_->publish(mask);
+
+    subscription_ = this->create_subscription<nav_msgs::msg::OccupancyGrid>(
+      MASK_TOPIC,
+      rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+      std::bind(&MaskPublisher::mapCallback, this, std::placeholders::_1));
+  }
+
+  ~MaskPublisher()
+  {
+    subscription_.reset();
+    publisher_.reset();
+    delete access_;
+  }
+
+  bool isPublished()
+  {
+    std::lock_guard<mutex_t> guard(*getMutex());
+    return is_published_;
+  }
+
+  mutex_t * getMutex()
+  {
+    return access_;
+  }
+
+private:
+  rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr publisher_;
+  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr subscription_;
+
+  void mapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr)
+  {
+    is_published_ = true;
+  }
+
+  bool is_published_;
+
+  mutex_t * access_;
+};  // MaskPublisher
+
+struct Point
+{
+  unsigned int x, y;
+};
+
+class TestNode : public ::testing::Test
+{
+public:
+  TestNode() {}
+
+  ~TestNode() {}
+
+protected:
+  void createMaps(unsigned char master_value, int8_t mask_value);
+  void publishMaps();
+  void createKeepoutFilter();
+  void reset();
+  void testKeepoutFilter(unsigned char free_value, unsigned char keepout_value);
+
+  std::shared_ptr<nav2_costmap_2d::KeepoutFilter> keepout_filter_;
+  std::shared_ptr<nav2_costmap_2d::Costmap2D> master_grid_;
+
+  std::vector<Point> keepout_points_;
+
+private:
+  void verifyMasterGrid(unsigned char free_value, unsigned char keepout_value);
+
+  nav2_util::LifecycleNode::SharedPtr node_;
+
+  std::shared_ptr<nav_msgs::msg::OccupancyGrid> mask_;
+
+  std::shared_ptr<InfoPublisher> info_publisher_;
+  std::shared_ptr<MaskPublisher> mask_publisher_;
+};
+
+void TestNode::createMaps(unsigned char master_value, int8_t mask_value)
```

Comment:
Why the master and mask values? You create it with `master_value` for all cells as background then just rewrite them all with `mask_value` when you resize. Either way, they're all identical.

Reply:
Nor sure, what do you mean here?
- `master_value` is used for filling `master_grid_` when creating. `master_grid_` is a `nav2_costmap_2d::Costmap2D` type object, so I am using just a provided constructor for it.
- `mask_value` is used for filling another object - `mask_` which is a `nav_msgs::msg::OccupancyGrid` type. For this type I did not meet a constructor like `OccupancyGrid(width, height, resolution, size, default_value)`, so I am making all this things manually in the code below.